x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskTests.java

```diff
@@ -218,23 +626,31 @@ public void markAsCompleted() {
 
             @Override
             public void markAsFailed(Exception e) {
+                fatalError = e;
                 stopped.set(true);
-                failureHolder.set(e);
             }
         };
     }
 
-    @After
-    public void cancelNodeTask() throws Exception {
-        if (task != null){
-            task.markAsCompleted();
-            assertThat(failureHolder.get(), nullValue());
-            assertBusy(() -> {
-                ShardFollowNodeTask.Status status = task.getStatus();
-                assertThat(status.getNumberOfConcurrentReads(), equalTo(0));
-                assertThat(status.getNumberOfConcurrentWrites(), equalTo(0));
-            });
+    private static ShardChangesAction.Response generateShardChangesResponse(long fromSeqNo, int size, long imdVersion,
+                                                                            long leaderGlobalCheckPoint) {
+        List<Translog.Operation> ops = new ArrayList<>();
+        for (long seqNo = fromSeqNo; seqNo < size; seqNo++) {
+            String id = UUIDs.randomBase64UUID();
+            byte[] source = "{}".getBytes(StandardCharsets.UTF_8);
+            ops.add(new Translog.Index("doc", id, seqNo, 0, source));
         }
+        return new ShardChangesAction.Response(imdVersion, leaderGlobalCheckPoint, ops.toArray(new Translog.Operation[0]));
     }
 
+    void startTask(ShardFollowNodeTask task, long leaderGlobalCheckpoint, long followerGlobalCheckpoint) {
+        task.start(followerGlobalCheckpoint);
+        // Shortcut to just set leaderGlobalCheckpoint, calling for example handleReadResponse() has side effects that
+        // complicates testing in isolation.
```

Comment:
I think we can have a ShardFollowNodeTask#start method that accepts two params: leaderGCP and followerGCP. Then we can remove updateLeaderGlobalCheckpoint and this helper method `startTask`. WDYT?

Reply:
Right, I think that in the case of this test that this would be better. The only downside is that in production code we will always set the leaderGCP and followerGCP to the same value.  ( The `ShardFollowTasksExecutor` only fetches the followerGCP). But I think this better then separating out updating the leaderGCP in a separate method.